[extractor/sbs] Overhaul extractor for new APIs

dirkf opened the pull request:
<details>
<summary>Boilerplate: own code, bug fix</summary>
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

</details>

---

### Description of your *pull request* and other information

Australian provider SBS has changed its hosting arrangements and APIs, breaking the existing extractor.

The PR is intended to deal with the new APIs.

Some specialisations of extractor methods are included:
* `_download_webpage_handle()` detects geo-restriction
* `_extract_m3u8_formats()` defaults to the native downloader.

As most shows of interest are geo-restricted to Australia, it needs thorough testing in-region (please).

The initial version retains the old extractor as dead code. If none of it turns out to be needed, a future commit will remove it.

Resolves #31841.
See also https://github.com/yt-dlp/yt-dlp/issues/6543.

gamer191 reviewed (commented):
Thanks heaps for this PR! I will test it tomorrow. Anything specific for me to test?

gamer191 commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,9 +78,191 @@ class SBSIE(InfoExtractor):
+            'uploader': 'SBSC',
+        }
+
+    def _old_real_extract(self, url):
```
```suggestion
    def _old_real_extract(self, url): //disabled
```

dirkf replied:
It'll be removed before merging.

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,9 +78,191 @@ class SBSIE(InfoExtractor):
+        'C': 7,
+        'G': 0,
+        'PG': 0,
+        'M': 14,
```
This was my source: https://rating-system.fandom.com/wiki/Australian_TV_content_rating_system

I had some trouble working out what it meant for M and MA15+. Definitely add NC.

Let's remove the `_` at the start of the class variable name and then it can be imported if required by one of the handful of other .au sites.

gamer191 replied:
> I had some trouble working out what it meant for M and MA15+. Are all these 15+?

M means it's recommended for people 15+
MA15+ means that it's strictly 15+ (cinemas ask for ID, for example)

> It doesn't mention X18+ either: == R18+ ?

X18+ is similar to R18+, but it's reserved for sexually explicit content (see https://www.classification.gov.au/classification-ratings/what-do-ratings-mean#X18+)


On a side note, it seems you were actually right, there is an MAV15+ category (see https://www.alrc.gov.au/publication/classification-content-regulation-and-convergent-media-alrc-report-118/9-classification-categories-and-criteria-2/classification-categories/)

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,9 +78,191 @@ class SBSIE(InfoExtractor):
+                'datePublished', ('publication', 'startDate'),
+                expected_type=parse_iso8601),
+            'channel': traverse_media(('taxonomy', 'channel', 'name')),
+            'uploader': 'SBSC',
```
This matches what was previously being returned (from ThePlatform?) but isn't seen anywhere in the metadata that we examined in the issue.

gamer191 replied:
It's from this weird commit 79ba9140dc8fcf5883b7473596e8f20cba6b479f

dirkf replied:
I presume it was added because that was what was returned by ThePlatorm (ie SBS Corp).

gamer191 commented on youtube_dl/extractor/sbs.py:
```diff
@@ -44,12 +44,13 @@ class SBSIE(InfoExtractor):
 
     _TESTS = [{
-        # Original URL is handled by the generic IE which finds the iframe:
+        # Exceptional unrestricted show for testing, thanks SBS,
```
What is this comment supposed to mean?

dirkf replied:
The test URL was found in an `<iframe>` of a page that no longer exists.

gamer191 commented on youtube_dl/extractor/sbs.py:
```diff
@@ -44,12 +44,13 @@ class SBSIE(InfoExtractor):
         'md5': 'e49d0290cb4f40d893b8dfe760dce6b0',
         'info_dict': {
-            'id': '320403011771',  # '_rFBPRPO4pMR',
+            'id': '320403011771',  # formerly '_rFBPRPO4pMR', no longer found
```
That url still shows this id on yt-dlp  (and downloads correctly). I will test tomorrow to see if it's a potential regression with this pr

dirkf replied:
As it's an old show, it's still available via ThePlatform, but its ID there doesn't seem to have survived being imported into the DB(s) underlying the new media APIs.

gamer191 commented on youtube_dl/extractor/sbs.py:
```diff
@@ -1,33 +1,66 @@
-        # Original URL is handled by the generic IE which finds the iframe:
+        # Exceptional unrestricted show for testing, thanks SBS,
+        # from an iframe of this page, handled by the generic IE, now 404:
         # http://www.sbs.com.au/thefeed/blog/2014/08/21/dingo-conservation
```
Is this where the url moved? https://www.sbs.com.au/programs/video/320403011771/Dingo-Conservation-The-Feed

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,9 +79,197 @@ class SBSIE(InfoExtractor):
+
+        return {
+            'id': video_id,
+            'title': traverse_media(('displayTitles', Ellipsis, 'title'),
+                                    get_all=False) or media['title'],
```
The extensive test suite (2 shows!) that I exercised included 1 show that was an episode, which had no specific episode name (`Episode 6`, or some such).

If the `title` or `episode` should be set to something like `{series} S{season_number}E{episode_number}` with any non-trivial `episode` value appended, that's fine. Show some examples.

icenov commented:
Just tried out the #31880 for new SBS platform and it worked fine. Bit of a learning experience for me as first time with git, but good docs helped! Thanks!

Hazza117 commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,9 +79,197 @@ class SBSIE(InfoExtractor):
+    def _get_player_data(self, video_id, headers=None, fatal=False):
+        return self._download_json(update_url_query(
+            self._PLAYER_API + 'video_stream', {'id': video_id, 'context': 'tv'}),
+            video_id, headers=headers, fatal=fatal) or {}
```
I believe `headers` should default to an empty dict here and in `_call_api`. Setting `headers` to None results in options like `--geo-bypass-country` breaking.

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,9 +79,197 @@ class SBSIE(InfoExtractor):
+            else:
+                kwargs['expected_status'] = exp
+                kwargs = compat_kwargs(kwargs)
+
```
Address https://github.com/ytdl-org/youtube-dl/pull/31880#discussion_r1153294487 (not required for yt-dlp or future yt-dl)
```suggestion

        # for now, ensure that headers passed is not None
        if len(args) > 5:
            if args[5] is None:
                # replace with empty dict
                args = list(args)
                args[5] = {}
        else:
            if kwargs.get('headers', False) is None:
                # enforce callee default
                del kwargs['headers']

```

Rygle commented:
Works great. Just copied and pasted the latest version from the pull request and it works. Thanks!

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -2,32 +2,61 @@
-        # Original URL is handled by the generic IE which finds the iframe:
+        # Exceptional unrestricted show for testing, thanks SBS,
+        # from an iframe of this page, handled by the generic IE, now 404:
         # http://www.sbs.com.au/thefeed/blog/2014/08/21/dingo-conservation
```
```suggestion
        # https://www.sbs.com.au/programs/video/320403011771/Dingo-Conservation-The-Feed
```

peternewell commented:
I have used this file to download a movie from SBS that previously I wasn't able to, so it appears to work ok.

Thanks for the change.

crazee1234 commented:
Code seems to download episodes OK, but not their subtitles (using --write-subs). Links to some episodes which have subtitles follow. https://www.sbs.com.au/ondemand/tv-series/alex-polizzis-secret-italy/season-1/alex-polizzi-secret-italy-s1-ep3/1378014275543 https://www.sbs.com.au/ondemand/tv-series/beerland/season-1/beerland-s1-ep1/1068730947696

Rygle commented:
> Code seems to download episodes OK, but not their subtitles

Same, no subtitles on several items I checked using yt-dlp, which is essentially this code afaik.

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,33 +77,141 @@ class SBSIE(InfoExtractor):
+            'season_number': (('partOfSeries', None), 'seasonNumber', T(int_or_none)),
+            'episode': ('epName', T(str)),
+            'episode_number': ('episodeNumber', T(int_or_none)),
+            'timestamp': ('datePublished', ('publication', 'startDate'), T(parse_iso8601)),
```
```suggestion
            'timestamp': (('datePublished', ('publication', 'startDate')), T(parse_iso8601)),
```

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,33 +77,141 @@ class SBSIE(InfoExtractor):
+            'age_limit': ('classificationID', 'contentRating',
+                          T(lambda x: self.AUS_TV_PARENTAL_GUIDELINES.get(x, '').upper() or None)),  # dict.get is unhashable in py3.7
+        }, get_all=False), traverse_obj(media, {
+            'categories': (('genres', Ellipsis), ('taxonomy', ('genre', 'subgenre'),
+                           'name', T(str))),
```
```suggestion
            'categories': ((('genres', Ellipsis),
                            ('taxonomy', ('genre', 'subgenre'), Ellipsis, 'name')),
                           T(str)),
```

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -2,32 +2,63 @@
             'timestamp': 1408613220,
             'upload_date': '20140821',
             'uploader': 'SBSC',
+            'tags': None,
+            'categories': None,
```
```suggestion
            'tags': 'mincount:10',
            'categories': 'count:2',
```

dirkf commented on youtube_dl/extractor/sbs.py:
```diff
@@ -46,33 +77,141 @@ class SBSIE(InfoExtractor):
+            'age_limit': ('classificationID', 'contentRating',
+                          T(lambda x: self.AUS_TV_PARENTAL_GUIDELINES.get(x, '').upper() or None)),  # dict.get is unhashable in py3.7
+        }, get_all=False), traverse_obj(media, {
+            'categories': (('genres', Ellipsis), ('taxonomy', ('genre', 'subgenre'),
+                           'name', T(str))),
+            'tags': (('consumerAdviceTexts', ('sbsSubCertification', 'consumerAdvice')),
+                     Ellipsis, T(str)),
```
```suggestion
            'categories': ((('genres', Ellipsis),
                            ('taxonomy', ((('genre', 'subgenre'), Ellipsis, 'name'), 'useType'))),
                           T(str)),
            'tags': ((((('keywords',),
                        ('consumerAdviceTexts', ('sbsSubCertification', 'consumerAdvice'))),
                       Ellipsis),
                     ('taxonomy', ('era', 'location', 'section', 'subject', 'theme'),
                      Ellipsis, 'name')),
                     T(str)),
```

dirkf commented:
I've updated the extractor from the latest _yt-dlp_ version and fixed the extraction of `categories` and `tags`. The Dingo Conservation test clip succeeds.

Live testing welcome.

peternewell commented:
I have inserted this sbs.py file into Youtube_Dl and it appears to work ok.

dirkf commented:
It's better than the master version, then. Good to go, I think.
